### DISCUSSION TITLE
An anecdote... 😏

### DISCUSSION THREAD

**Discussion by benhutchison** [armanbilge/schrodinger#336] [Ideas]
Whoops, somehow using an emoji created the issue prematurely. 

Anyway, an anecdote..

In an unpublished project, I have opaque types that model non-negative Integers, Doubles and Decimals. 

Initially, I added bespoke operations as extension methods to each, but that got boring fast. So I switched to using Algebra; define surface-syntax in terms of the typeclasses in Typelevel algebra and Spire, and then implement those TCs for my non-negative types. Has proved quite practical, and feels like a much more solid base to go forward upon. 

However. It turned out to be far from obvious what algebras to implement for non-negative integral and rational data types. I got as far as `Semifield` which gave me everything but subtraction. Solving subtraction led me down a fascinating wikipedia rabbit hole and ultimately to the discovery of `Monus`.

But then I hit another doubt. What algebraic structure models non-negative rationals, in entirety? M-Semiring doesnt include division. Semifield doesn't include monus-subtraction. Is there such thing as a "M-Semifield"? 🤔 💭 

Surely the combined mathematical GDP of the world has been able to give name to the algebra of such a ubiquitous class of numbers as the non-negative rationals/reals..? surely..?

And it was in the spirit of doubtful inquiry that I accidentally [googled](https://www.google.com/search?q=semifield+monus) your project, number 2 after wikipedia. "Arman Bilge woz here!"

And sure enough, it does seem like the weight type `W` in your [Resampler](https://github.com/armanbilge/stratus/blob/main/core/src/main/scala/stratus/Resampler.scala#L51) is intended to be a non-negative rational.

And that felt like an anecdote worth sharing... :)

**Comment by benhutchison** [comment]
Actually, can I ask you about your take on `Monus`?

You extend `PartialOrder` which is consistent with wikipedia etc. However, IIUC, a partial ordering means that this Monus cannot be a Metric, because if two elements a & b are incomparable, we cant use eg `max(monus(a, b),  monus(b, a))` to find a distance between them. 

We can sensibly define distance between pair of non-negative numbers. In your view, is there a case for an OrderedMonus typeclass, ie Monus+Order, that could also implement eg Spire's `MetricSpace[N, N]`?

**Reply by armanbilge** [reply]
> However, IIUC, a partial ordering means that this Monus cannot be a Metric

Sorry to split hairs: it means that a Monus is not necessarily a Metric, but it does not mean it cannot be a Metric.

> In your view, is there a case for an OrderedMonus typeclass, ie Monus+Order

I need to load this stuff back into my brain. But at a glance, this seems reasonable: after all, there are examples of Monus with total Order.

**Comment by benhutchison** [comment]
Another question, if I may. 

Your Monus chooses a "has a" over trait inheritance "is a":
```
trait Monus[A]:

  def monus(x: A, y: A): A

  def additiveCommutativeMonoid: AdditiveCommutativeMonoid[A]
  def naturalOrder: PartialOrder[A]
```

Typelevel Algebra generally prefers the inheritance approach (as do I, I think) eg [Field](https://github.com/typelevel/cats/blob/main/algebra-core/src/main/scala/algebra/ring/Field.scala), but it is not 100% consistent see eg [Signed](https://github.com/typelevel/cats/blob/main/algebra-core/src/main/scala/algebra/ring/Signed.scala#L53)

Is there a design heuristic here that you follow and can express, or is this simply a case of different practices colliding and a lack of consensus?

**Reply by armanbilge** [reply]
> Is there a design heuristic here that you follow and can express, or is this simply a case of different practices colliding and a lack of consensus?

Yes, actually there is a blog post on this exact topic. It's also briefly explained in the comment on [`Signed`](https://github.com/typelevel/cats/blob/32cf51f3dc3206a2f1e86671e904567eb89e3b65/algebra-core/src/main/scala/algebra/ring/Signed.scala#L48-L49).

https://typelevel.org/blog/2016/09/30/subtype-typeclasses.html

The tl;dr is that the inheritance encoding only works for a "blessed" linear hierarchy, because otherwise you may easily end up with more than one implicit instance satisfying the same constraint, leading to ambiguity.

**Reply by benhutchison** [reply]
Right. I do recall this issue now, although back of mind as I haven't been bitten recently.

So, using the Resampler signature as a concrete example to align on the problem:
```
 def targetWeight[F[_]: Monad, W: Monus: Order](computeTarget: Eagle[W] => F[W])(using
      W: Semifield[W],
      cat: Categorical[F, NonEmptyList[W], Long],
      unif: DiscreteUniform[F, Long],
  ): Resampler[F, W]
```

If `Monus` inherited `AdditiveCommutativeMonoid`, it would become ambiguous WRT `Semifield` if a method called within the body needed eg `AdditiveCommutativeMonoid[W]`, since they both extend it. Diamond problem again. 😬 


It's awkward. I watched @adelbertc's talk, the dotty [issue](https://github.com/lampepfl/dotty-feature-requests/issues/5), lots of problem, not much solution there.. 


I have a proposed solution that I believe is workable (may require a time-machine 😝). To sketch it, let me make up some terms: _bare_ vs _bundled_ typeclasses (TC)

[Monus](https://github.com/armanbilge/schrodinger/blob/be55d296c411f79564e70f56409480e6c9d5c6b7/math/shared/src/main/scala/schrodinger/math/Monus.scala) is an example of a _bare_ TC:
- Doesnt inherit from other TCs, declares dependencies as members instead
- Interface defines new operations only, in terms of the field
- Bare TCs wont cause ambiguity
- Design challenges is that clients using _bare_ TCs need to express lots of fine-grained constraints, eg `W: Monus: Order: Semifield` above.

OTOH [Field](https://github.com/typelevel/cats/blob/main/algebra-core/src/main/scala/algebra/ring/Field.scala) is an example of a _bundled_ TC:
- Inherits from all its dependencies and extends their Interface
- Bundles are often convenient for clients because they "bundle up" constraints, eg a hypothetical `W: OrderedMSemifield` 
- Bundled TCs can cause ambiguity when 2+ bundles share a common ancestor

Proposed design heuristic:
- All TCs must be defined bare form. 
- The bare TCs can be optionally bundled for convenience. Bundles should be declared for "commonly needed" combinations - this is a judgement call involving taste & style.
- There is a naming convention for distinguishing bare vs bundle TCs
- Clients can always safely depend upon [0-1] bundle TCs plus any number of bare TCs. Of course, sometimes, 2+ bundle TCs from non-overlapping domains will be fine.

I know the existing corpus of Scala/Typelevel TCs are not declared in bare form, but I still feel it is important to have a view on what one _ought_ to do.. 🧭

**Reply by benhutchison** [reply]
Implementing the above conventions would see bundle instances and bare instances sharing the same implicit scope. 

I become concerned that this would be ambiguous and thus an error, but empirically it [doesn't seem to be the case](https://users.scala-lang.org/t/why-is-this-not-an-ambiguous-implicit/9623).

**Reply by benhutchison** [reply]
There is an attempt to address this problem in Scala 3.6/2.7:

https://www.scala-lang.org/2024/08/19/given-priority-change-3.7.html

Will be interesting to see how it works in practice on larger typeclass-heavy Scala codebases :)

**Comment by armanbilge** [comment]
(Moved to a discussion on this repo, where [`Monus`](https://github.com/armanbilge/schrodinger/blob/be55d296c411f79564e70f56409480e6c9d5c6b7/math/shared/src/main/scala/schrodinger/math/Monus.scala) is defined as well as an implementation [`LogDouble`](https://github.com/armanbilge/schrodinger/blob/be55d296c411f79564e70f56409480e6c9d5c6b7/math/shared/src/main/scala/schrodinger/math/LogDouble.scala).)

> M-Semiring doesnt include division. Semifield doesn't include monus-subtraction. Is there such thing as a "M-Semifield"?

I actually haven't heard of the "M-Semiring", can you link me anything about that?

**Reply by benhutchison** [reply]
I picked it up [here](https://en.wikipedia.org/wiki/Monus#Other_structures) and ran with it... 😝
